Source/Core/VideoCommon/BPMemory.h

```diff
@@ -354,169 +469,384 @@ struct TevStageCombiner
   u32 getTexMap(int i) const { return i ? texmap1.Value() : texmap0.Value(); }
   u32 getTexCoord(int i) const { return i ? texcoord1.Value() : texcoord0.Value(); }
   u32 getEnable(int i) const { return i ? enable1.Value() : enable0.Value(); }
-  u32 getColorChan(int i) const { return i ? colorchan1.Value() : colorchan0.Value(); }
+  RasColorChan getColorChan(int i) const { return i ? colorchan1.Value() : colorchan0.Value(); }
+};
+template <>
+struct fmt::formatter<TwoTevStageOrders>
+{
+  constexpr auto parse(format_parse_context& ctx) { return ctx.begin(); }
+  template <typename FormatContext>
+  auto format(const TwoTevStageOrders& stages, FormatContext& ctx)
+  {
+    return format_to(ctx.out(),
+                     "Stage 0 texmap: {}\nStage 0 tex coord: {}\n"
+                     "Stage 0 enable texmap: {}\nStage 0 color channel: {}\n"
+                     "Stage 1 texmap: {}\nStage 1 tex coord: {}\n"
+                     "Stage 1 enable texmap: {}\nStage 1 color channel: {}\n",
+                     stages.texmap0, stages.texcoord0, stages.enable0 ? "Yes" : "No",
+                     stages.colorchan0, stages.texmap1, stages.texcoord1,
+                     stages.enable1 ? "Yes" : "No", stages.colorchan1);
+  }
 };
 
 union TEXSCALE
 {
-  struct
-  {
-    u32 ss0 : 4;  // Indirect tex stage 0, 2^(-ss0)
-    u32 ts0 : 4;  // Indirect tex stage 0
-    u32 ss1 : 4;  // Indirect tex stage 1
-    u32 ts1 : 4;  // Indirect tex stage 1
-    u32 pad : 8;
-    u32 rid : 8;
-  };
+  BitField<0, 4, u32> ss0;   // Indirect tex stage 0, 2^(-ss0)
+  BitField<4, 4, u32> ts0;   // Indirect tex stage 0
+  BitField<8, 4, u32> ss1;   // Indirect tex stage 1
+  BitField<12, 4, u32> ts1;  // Indirect tex stage 1
   u32 hex;
 };
+template <>
+struct fmt::formatter<TEXSCALE>
+{
+  constexpr auto parse(format_parse_context& ctx) { return ctx.begin(); }
+  template <typename FormatContext>
+  auto format(const TEXSCALE& scale, FormatContext& ctx)
+  {
+    return format_to(ctx.out(),
+                     "Even stage S scale: {} ({})\n"
+                     "Even stage T scale: {} ({})\n"
+                     "Odd stage S scale: {} ({})\n"
+                     "Odd stage T scale: {} ({})",
+                     scale.ss0, 1.f / (1 << scale.ss0), scale.ts0, 1.f / (1 << scale.ts0),
+                     scale.ss1, 1.f / (1 << scale.ss1), scale.ts1, 1.f / (1 << scale.ts1));
+  }
+};
 
 union RAS1_IREF
 {
-  struct
-  {
-    u32 bi0 : 3;  // Indirect tex stage 0 ntexmap
-    u32 bc0 : 3;  // Indirect tex stage 0 ntexcoord
-    u32 bi1 : 3;
-    u32 bc1 : 3;
-    u32 bi2 : 3;
-    u32 bc3 : 3;
-    u32 bi4 : 3;
-    u32 bc4 : 3;
-    u32 rid : 8;
-  };
+  BitField<0, 3, u32> bi0;  // Indirect tex stage 0 ntexmap
+  BitField<3, 3, u32> bc0;  // Indirect tex stage 0 ntexmap
+  BitField<6, 3, u32> bi1;
+  BitField<9, 3, u32> bc1;
+  BitField<12, 3, u32> bi2;
+  BitField<15, 3, u32> bc3;  // Typo?
+  BitField<18, 3, u32> bi4;
+  BitField<21, 3, u32> bc4;
```

Comment:
These names are somewhat odd since it skips `bc2` and `bi3`; I don't think it should have gone up to 4 at all.  I've left this unchanged in case there's some intended meaning though.